### PR TITLE
Convert right_panel to TS

### PR DESCRIPTION
**Pull request by swapnilraj**
Converts right_panel stuff to TS. 
I have left out `right_panel/UserInfo.js` for now, as the PR is a bit big.

Signed-off-by: Swapnil Raj sr@sraj.me

### CODE REVIEW

**Comment by JorikSchellekens** [comment]
Reached out to swapnilraj and found the problem with the tests most likely to be 
```
Uncaught TypeError: Cannot read property 'userId' of undefined
    at pendingVerificationRequestForUser (verification.js?ee94:147)
    at RightPanelStore.__onDispatch (RightPanelStore.ts?8df7:160)
    at RightPanelStore.__invokeOnDispatch (FluxStore.js?5294:148)
    at Object.eval [as ID_23] (FluxStore.js?5294:79)
    at MatrixDispatcher._invokeCallback (Dispatcher.js?1f1c:198)
    at MatrixDispatcher.dispatch (Dispatcher.js?1f1c:174)
```

**Comment by swapnilraj** [comment]
> Reached out to swapnilraj and found the problem with the tests most likely to be

Thanks for your help on this, I think this PR is ready to reviewed now.

File: src/components/views/right_panel/HeaderButton.tsx [hunk 1]
```diff
@@ -19,18 +19,33 @@ limitations under the License.
...
+}
+
+export default class HeaderButton extends React.Component<IProps> {
+    constructor(props: IProps) {
+        super(props);
         this.onClick = this.onClick.bind(this);
     }
 
-    onClick(ev) {
+    onClick(_ev: React.KeyboardEvent) {
```
Comment (JorikSchellekens) [inline src/components/views/right_panel/HeaderButton.tsx:48]: We don't really do the '_' thing

File: src/components/views/right_panel/RoomHeaderButtons.tsx [hunk 2]
```diff
@@ -21,82 +21,82 @@ limitations under the License.
...
     }
 
-    _onNotificationsClicked() {
+    private onNotificationsClicked() {
         // This toggles for us, if needed
-        this.setPhase(RIGHT_PANEL_PHASES.NotificationPanel);
+        this.setPhase(RightPanelPhases.NotificationPanel);
     }
 
     renderButtons() {
```
Comment (JorikSchellekens) [inline src/components/views/right_panel/RoomHeaderButtons.tsx:82]: public please

File: src/components/views/right_panel/RoomHeaderButtons.tsx [hunk 3]
```diff
@@ -21,82 +21,82 @@ limitations under the License.
...
-        this._onMembersClicked = this._onMembersClicked.bind(this);
-        this._onFilesClicked = this._onFilesClicked.bind(this);
-        this._onNotificationsClicked = this._onNotificationsClicked.bind(this);
+        super(props, HeaderKind.Room);
+        this.onMembersClicked = this.onMembersClicked.bind(this);
+        this.onFilesClicked = this.onFilesClicked.bind(this);
+        this.onNotificationsClicked = this.onNotificationsClicked.bind(this);
     }
 
     onAction(payload: ActionPayload) {
```
Comment (JorikSchellekens) [inline src/components/views/right_panel/RoomHeaderButtons.tsx:44]: private

File: src/components/views/right_panel/HeaderButtons.tsx [hunk 4]
```diff
@@ -21,42 +21,54 @@ limitations under the License.
...
-
         const rps = RightPanelStore.getSharedInstance();
         this.state = {
             headerKind: kind,
-            phase: kind === HEADER_KIND_ROOM ? rps.visibleRoomPanelPhase : rps.visibleGroupPanelPhase,
+            phase: kind === HeaderKind.Room ? rps.visibleRoomPanelPhase : rps.visibleGroupPanelPhase,
         };
     }
 
     componentDidMount() {
```
Comment (JorikSchellekens) [inline src/components/views/right_panel/HeaderButtons.tsx:55]: public on all of these please

File: src/components/views/right_panel/HeaderButton.tsx (hunk 1 above)
Comment (JorikSchellekens) [reply src/components/views/right_panel/HeaderButton.tsx:48]: It's not important though

Comment (swapnilraj) [reply src/components/views/right_panel/HeaderButton.tsx:48]: The `ev` parameter is not being used, so I wanted to make that explicit.

Comment (t3chguy) [reply src/components/views/right_panel/HeaderButton.tsx:48]: You could just omit the parameter

Comment (swapnilraj) [reply src/components/views/right_panel/HeaderButton.tsx]: I didn't want to change the type signature, but it seems that's alright to do. Done

File: src/components/views/right_panel/RoomHeaderButtons.tsx (hunk 3 above)
Comment (swapnilraj) [reply src/components/views/right_panel/RoomHeaderButtons.tsx]: This should be public since the subclass uses it.

Comment (t3chguy) [reply src/components/views/right_panel/RoomHeaderButtons.tsx]: if a subclass is using it then shouldn't it be `protected`?

Comment (swapnilraj) [reply src/components/views/right_panel/RoomHeaderButtons.tsx]: Oops should have paid more attention in those OOP classes.

**Pull request merged**